components/fonts/shapers/mod.rs

```diff
@@ -0,0 +1,237 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+mod harfbuzz;
+use std::cmp;
+
+use app_units::Au;
+use base::text::is_bidi_control;
+use euclid::default::Point2D;
+use fonts_traits::ByteIndex;
+pub use harfbuzz::{ShapedGlyphData, Shaper};
+use log::debug;
+use num_traits::Zero as _;
+
+const NO_GLYPH: i32 = -1;
+
+use crate::{
+    Font, FontBaseline, GlyphData, GlyphId, GlyphStore, ShapingOptions, advance_for_shaped_glyph,
+};
+
+/// Utility function to convert a `unicode_script::Script` enum into the corresponding `c_uint` tag that
+/// harfbuzz uses to represent unicode scipts.
+fn unicode_script_to_iso15924_tag(script: unicode_script::Script) -> u32 {
+    let bytes: [u8; 4] = match script {
+        unicode_script::Script::Unknown => *b"Zzzz",
+        _ => {
+            let short_name = script.short_name();
+            short_name.as_bytes().try_into().unwrap()
+        },
+    };
+
+    u32::from_be_bytes(bytes)
+}
+
+struct ShapedGlyphEntry {
+    codepoint: GlyphId,
+    advance: Au,
+    offset: Option<Point2D<Au>>,
+}
+
+/// A servo shaper
+pub(crate) trait TShaper: Send + Sync {
+    /// Calculate the layout metrics associated with the given text when painted in a specific font.
+    fn shape_text(&self, text: &str, options: &ShapingOptions, glyphs: &mut GlyphStore);
+    fn baseline(&self) -> Option<FontBaseline>;
+}
+
+/// Holds the results of shaping. Abstracts over HarfBuzz and HarfRust which return data in very similar
+/// form but with different types
+trait THarfShapedGlyphData {
+    /// The number of shaped glyphs
+    fn len(&self) -> usize;
+    /// The byte offset of the shaped glyph in the souce text
+    fn byte_offset_of_glyph(&self, i: usize) -> u32;
+    /// Returns shaped glyph data for one glyph, and updates the y-position of the pen.
+    fn entry_for_glyph(&self, i: usize, y_pos: &mut Au) -> ShapedGlyphEntry;
+}
+
+/// A Servo shaper based on HarfBuzz (HarfBuzz or HarfRust)
+trait THarfShaper: Send + Sync {
```

Comment:
I don't think we typically use the `T` prefix in Servo, so please rename before landing. I would also avoid the "Harf" terminology as it's a bit obscure unless you know this is dealing with two libraries, one called HarfBuzz and the other called HarfRust.

```suggestion
trait HarfBuzzShaperTrait: Send + Sync {
```

Reply:
Stylo uses this naming scheme. But happy to change it :)